pkg/router/apisix.go

```diff
@@ -0,0 +1,215 @@
+/*
+Copyright 2022 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package router
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	a6v2 "github.com/fluxcd/flagger/pkg/apis/apisix/v2"
+	flaggerv1 "github.com/fluxcd/flagger/pkg/apis/flagger/v1beta1"
+	clientset "github.com/fluxcd/flagger/pkg/client/clientset/versioned"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// ApisixRouter is managing Apisix Route
+type ApisixRouter struct {
+	apisixClient clientset.Interface
+	logger       *zap.SugaredLogger
+	setOwnerRefs bool
+}
+
+const maxPriority = 10000
+
+// Reconcile creates or updates the Apisix Route
+func (ar *ApisixRouter) Reconcile(canary *flaggerv1.Canary) error {
+	if canary.Spec.RouteRef == nil || canary.Spec.RouteRef.Name == "" {
+		return fmt.Errorf("apisix route selector is empty")
+	}
+
+	apisixRoute, err := ar.apisixClient.ApisixV2().ApisixRoutes(canary.Namespace).Get(context.TODO(), canary.Spec.RouteRef.Name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("apisix route %s.%s get query error: %w", canary.Spec.RouteRef.Name, canary.Namespace, err)
+	}
+
+	apisixRouteClone := apisixRoute.DeepCopy()
+	if len(apisixRouteClone.Spec.HTTP) != 1 && len(apisixRouteClone.Spec.HTTP[0].Backends) != 1 {
+		return fmt.Errorf("apisix route %s.%s only one http backend is supported", canary.Spec.RouteRef.Name, canary.Namespace)
+	}
+
+	httpBackend := apisixRouteClone.Spec.HTTP[0]
+	httpBackend.Priority = maxPriority
+
+	_, primaryName, canaryName := canary.GetServiceNames()
+
+	primaryBackend := httpBackend.Backends[0]
+	primaryBackend.ServiceName = primaryName
+	primaryWeight := 100
+	primaryBackend.Weight = &primaryWeight
+	httpBackend.Backends[0] = primaryBackend
+
+	canaryWeight := 0
+	canaryBackend := a6v2.ApisixRouteHTTPBackend{
+		ServiceName:        canaryName,
+		ServicePort:        primaryBackend.ServicePort,
+		ResolveGranularity: primaryBackend.ResolveGranularity,
+		Weight:             &canaryWeight,
+		Subset:             primaryBackend.Subset,
+	}
+	httpBackend.Backends = append(httpBackend.Backends, canaryBackend)
+
+	apisixRouteClone.Spec.HTTP[0] = httpBackend
+	canaryApisixRouteName := fmt.Sprintf("%s-canary", canary.Spec.RouteRef.Name)
```

Comment:
```suggestion
	canaryApisixRouteName := fmt.Sprintf("%s-%s-canary", canary.Spec.RouteRef.Name, httpBackend.Name)
```

since we are looping through the entire `ApisixRoute` object, a Canary might target different backends present in the same `ApisixRoute` object